Kernel/RobSubstitution.hpp

```diff
@@ -17,55 +17,331 @@
 #define __RobSubstitution__
 
 #include "Forwards.hpp"
+#include "Kernel/Polynomial.hpp"
 #include "Lib/Backtrackable.hpp"
+#include "Lib/Recycled.hpp"
 #include "Term.hpp"
-#include "MismatchHandler.hpp"
 #include "Lib/Hash.hpp"
 #include "Lib/DHMap.hpp"
+#include "Lib/Metaiterators.hpp"
+#include "Kernel/BottomUpEvaluation.hpp"
+#include "Lib/Environment.hpp"
+#include "Kernel/Signature.hpp"
+#include "Kernel/TypedTermList.hpp"
+#include <initializer_list>
 
 #if VDEBUG
 #include <iostream>
 #include "Lib/VString.hpp"
 #endif
 
+
+const int GLUE_INDEX=-3;
+const int SPECIAL_INDEX=-2;
+const int UNBOUND_INDEX=-1;
 namespace Kernel
 {
+struct VarSpec
+{
+  VarSpec() {}
+  VarSpec(unsigned var, int index) : var(var), index(index) {}
+
+  friend std::ostream& operator<<(std::ostream& out, VarSpec const& self);
+
+  /** number of variable */
+  unsigned var;
+  /** index of variable bank */
+  int index;
+
+  auto asTuple() const { return std::tie(var, index); }
```

Comment:
`std::tie` doesn't copy but creates references, so if you'd call `std::tie(_myStack, _myInt)` it will be of type `std::tuple<Stack<T>&, int&>` instead of copying the stack.

Reply:
I just always use `std::tie` for the `asTuple` functions so it's future proof if someone adds a field that is not trivially copy constructible.